qualtran/bloqs/mcmt/ctrl_spec_and.py

```diff
@@ -224,3 +226,111 @@ def _ctrl_on_multiple_values() -> CtrlSpecAnd:
     bloq_cls=CtrlSpecAnd,
     examples=(_ctrl_on_int, _ctrl_on_bits, _ctrl_on_nd_bits, _ctrl_on_multiple_values),
 )
+
+
+@frozen
+class ControlledViaAnd(Bloq):
+    """Reduces a generic control bloq to a singly-controlled bloq using an And ladder.
+
+    Implements a generic controlled version of the subbloq, by first reducing the
+    arbitrary control to a single qubit, and then using a single-qubit-controlled
+    variant of the subbloq.
+
+    For signature, see :class:`Controlled`.
+
+    Args:
+        subbloq: The bloq we are controlling.
+        ctrl_spec: The specification for how to control the bloq.
+    """
+
+    subbloq: Bloq
+    ctrl_spec: CtrlSpec
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return self._signature_and_ctrl_regs_from_controlled[0]
+
+    @cached_property
+    def ctrl_reg_names(self) -> tuple[str, ...]:
+        return self._signature_and_ctrl_regs_from_controlled[1]
+
+    @cached_property
+    def _signature_and_ctrl_regs_from_controlled(self) -> tuple[Signature, tuple[str, ...]]:
+        cbloq = Controlled(self.subbloq, self.ctrl_spec)
+        return cbloq.signature, tuple(cbloq.ctrl_reg_names)
```

Comment:
I think it's worth pulling out the logic in `Controlled.ctrl_reg_names` and `Controlled.ctrl_regs` into a free function `def ctrl_regs_from_ctrl_spec(ctrl_spec: CtrlSpec, subbloq: Optional[Bloq] = None)` which will return you a list of ctrl registers given a ctrl spec and subbloq. Then we won't need to allocate the `Controlled` subbloq only to define the signature of this bloq.